pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotUpsertCapacityEstimationRestletResource.java

```diff
@@ -0,0 +1,162 @@
+      @ApiParam(value = "numPartitions in string format") @QueryParam("numPartitions") String numPartitionsStr) {
+    ObjectNode resultData = JsonUtils.newObjectNode();
+    TableAndSchemaConfig tableSchemaConfig;
+
+    try {
+      tableSchemaConfig = JsonUtils.stringToObject(tableSchemaConfigStr, TableAndSchemaConfig.class);
+    } catch (IOException e) {
+      throw new ControllerApplicationException(LOGGER,
+          String.format("Invalid TableSchemaConfigs json string: %s", tableSchemaConfigStr),
+          Response.Status.BAD_REQUEST, e);
+    }
+
+    TableConfig tableConfig = tableSchemaConfig.getTableConfig();
+    resultData.put("tableName", tableConfig.getTableName());
+
+    Schema schema = tableSchemaConfig.getSchema();
+
+    // Estimated key space, it contains primary key columns
+    int bytesPerKey = 0;
+    List<String> primaryKeys = schema.getPrimaryKeyColumns();
+
+    if (primaryKeySize != null) {
+      bytesPerKey += Integer.valueOf(primaryKeySize);
+    } else {
+      for (String primaryKey : primaryKeys) {
+        FieldSpec.DataType dt = schema.getFieldSpecFor(primaryKey).getDataType();
+        if (dt == FieldSpec.DataType.JSON || dt == FieldSpec.DataType.LIST || dt == FieldSpec.DataType.MAP) {
+          String msg = "Not support data types for primary key columns";
+          throw new ControllerApplicationException(LOGGER, msg, Response.Status.BAD_REQUEST);
+        } else if (dt == FieldSpec.DataType.STRING) {
+          String msg = "Missing primary key sizes for String columns";
+          throw new ControllerApplicationException(LOGGER, msg, Response.Status.BAD_REQUEST);
+        } else {
+          bytesPerKey += dt.size();
+        }
+      }
+      // Java has a 24 bytes array overhead and there's also 8 bytes for the actual array object
+      bytesPerKey += 32;
+    }
+
+    // Estimated value space, it contains <segmentName, DocId, ComparisonValue(timestamp)> and overhead.
+    int bytesPerValue = 64;
+    String comparisonColumn = tableConfig.getUpsertConfig().getComparisonColumn();
+    if (comparisonColumn != null) {
+      FieldSpec.DataType dt = schema.getFieldSpecFor(comparisonColumn).getDataType();
+      if (dt == FieldSpec.DataType.STRING || dt == FieldSpec.DataType.JSON || dt == FieldSpec.DataType.LIST
+          || dt == FieldSpec.DataType.MAP) {
+        String msg = "Not support data types for the comparison column";
+        throw new ControllerApplicationException(LOGGER, msg, Response.Status.BAD_REQUEST);
+      } else {
+        bytesPerValue = 52 + dt.size();
+      }
+    }
+
+    resultData.put("bytesPerKey", bytesPerKey);
+    resultData.put("bytesPerValue", bytesPerValue);
+
+    long primaryKeyCardinality = Long.valueOf(cardinality);
+    long totalKeySpace = bytesPerKey * primaryKeyCardinality;
+    long totalValueSpace = bytesPerValue * primaryKeyCardinality;
+    long totalSpace = totalKeySpace + totalValueSpace;
```

Comment:
(optional) Here we only calculate the map content size. Should we also consider the map entry size and the array size within the map?

Reply:
The object overhead will vary between different machines. 

I found this https://dzone.com/articles/size-entry-map shows 81.4 bytes per concurrentHashmap entry, do you think this is accurate and want me to add that?

Reply:
In order to make more accurate estimation, we should count the object overhead because as you can see, the object overhead can be larger than the actual content within the entry.
I don't know the formula of object overhead from map size, so we need to do some research here. If you find that is too hard, we may add a TODO and address that later. But users should know this size does not include the overhead, so it can be quite off.

Reply:
Thanks for suggestion! I can add a Todo and address later. I will do some benchmark with heap dump.